### PR TITLE
Don't save form8 with large values in session

### LINKED ISSUE #230: Large values for form fields can trigger CookieOverflow (via saved forms)
It's possible to overflow the 4kb limit on cookie values because we're using `ActionDispatch::Session::CookieStore` for arbitrary user data in saved forms, 

We can

1. Skip saved form functionality if serialized form exceeds 4kb 
2. Use a different [session store](http://guides.rubyonrails.org/action_controller_overview.html#session)
3. Length-check all the fields, ensure that it's impossible to exceed 4kb...it may be that the user was trying enter a block of text that was too long to fit in form anyway.

Steps to reproduce:

1. Paste a large amount of text into the remarks field, 
2. Click certify

### PR DESCRIPTION
**Pull request by orischwartz**
Also adds maxlength to textarea fields so we don't accept values that can't fit in form8 anyway.

I suspect that guessing the serialized length via hash.to_s to too restrictive but this will at least prevent us from exceeding 4kb cookie limit. 

I've also tried GZipping/base64 encoding the values but that doesn't help. A better solution would be to override rails cookie store and gzip there.

Closes #230

### CODE REVIEW

**Comment by orischwartz** [comment]
@shanear  @paultag any better ideas for how to solve this?

File: app/controllers/certifications_controller.rb [hunk 1]
```diff
@@ -33,7 +33,16 @@ def new
 
   def create
     @form8 = Form8.new(params[:form8])
-    session[:form8] = @form8.attributes
+    serialized_form8 = @form8.attributes
+
+    # rough estimate
+    serialized_length = serialized_form8.to_s.bytesize
```
Comment (shanear) [inline app/controllers/certifications_controller.rb]: we shouldn't do this in the controller. but otherwise i think its fine

Comment (shanear) [inline app/controllers/certifications_controller.rb]: `form8.fits_in_session?` method maybe?

**Comment by KHarshawat** [comment]
@orischwartz @shanear if we're instituting a max size (and a char limit on the text field) we should definitely give users some feedback that this is happening. Can you display a character count in each of those boxes?

File: app/controllers/certifications_controller.rb (hunk 1 above)
Comment (orischwartz) [inline app/controllers/certifications_controller.rb]: I'm happy to make the change but:

Isn't the controller HTTP aware? And isn't the model protocol agnostic? Adding `fits_in_cookie `to form8 adds http-specific logic into a layer that shouldn't know about HTTP.

Comment (paultag) [inline app/controllers/certifications_controller.rb]: I think I'm with Ori on that, but I don't know much about Rails, but I'd expect Models to not worry about serialization size, IMO. Thoughts?

Comment (shanear) [inline app/controllers/certifications_controller.rb]: I think it's fine for models to have some knowledge of http. It's better than having a chunk of untested logic hanging out in a private controller function.

I think the most important thing is to keep controllers lean.

Comment (orischwartz) [inline app/controllers/certifications_controller.rb]: OK, I'll defer to you here but there are a couple of things I don't like about the new version--

* Form is serialized twice--once to check the length and again for serialization
* HTTP-specific logic is introduced into what was previously a pure model object.

On the plus side, the resulting controller is easier to read and more testable.

Comment (shanear) [inline app/controllers/certifications_controller.rb]: we could put the `fits_in_cookie?`method into a module or concern that we include into the controller. something like this:

```
module CookieStorage
  MAX_SIZE = 3.kilobytes

  def fits_in_cookie?(object)
    # ...
  end
end
```

```
attributes = @form8.attributes
session[:form8] = attributes if fits_in_cookie?(attributes)
```

Comment (orischwartz) [inline app/controllers/certifications_controller.rb]: Cool I like that. But we have users complaining so we'll deploy as is and add an issue for refactoring.

Comment (paultag) [inline app/controllers/certifications_controller.rb]: SGTM; let's file a bug to track that

**Pull request merged**